maintenance/code-backports.sh

```diff
@@ -0,0 +1,73 @@
+#!/bin/sh
+
+# silence debug output from VCS
+beQuiet="--quiet"
+#beQuiet=""
+version=`basename "$1" | sed s/squid-//`
+vnext=$(( $version + 1 ))
+srcbranch=v$vnext
+
+cd $1
+
+# Dev branch does not receive backports
+test -f .BASE && exit 0
+
+# clean the current workspace
+gitCleanWorkspace ()
+{
+  git clean $beQuiet -xdf --exclude="\.BASE"
+  git checkout $beQuiet -- .
+}
+
+if ! test -d .git ; then
+  echo "ERROR: missing git repository"
+  exit 1;
+fi
+if ! test -d ../squid-$vnext ; then
+  echo "ERROR: missing git repository to backport from"
+  exit 1;
+fi
+if test -e ../squid-$vnext/.BASE ; then
+  srcbranch=`cat ../squid-$vnext/.BASE`
+fi
+
+echo "FOUDN srcbranch=$srcbranch"
+
+# Prepare backports branch for updates
+git checkout $beQuiet v$version-next-backports 2>/dev/null &&
+  gitCleanWorkspace &&
+    git rebase origin/v$version 2>/dev/null || ( git rebase --abort 2>/dev/null ; exit 1 )
+
+# handler on successful backport
+doneCherryPick ()
+{
+  sha="$1"
+  prnum="$2"
+  version="$3"
+
+  git push $beQuiet >/dev/null
+  # TODO automatically remove label from PRs on successful commit+push to vN-next-backports
+}
+
+# Find backports to attempt:
+prlist=`gh pr list -L 1 --repo squid-cache/squid --state closed --label backport-to-v$version | wc -l`
+if test "$prlist" -ne 0; then
+  gh pr list --repo squid-cache/squid --state closed --label backport-to-v$version | while read prnum text; do
+    # find a commit in $srcbranch with " (#$prnum)" subject suffix
+    git log --oneline github/$srcbranch --grep=" (#$prnum)\$" | while read sha rest; do
+      git cherry-pick $sha && doneCherryPick $sha $prnum $version || cherry-pick --abort 2>/dev/null
+      gitCleanWorkspace
+    done
+  done
+
+  # Create a PR to merge (if needed)
+  if test `git diff github/$srcbranch v$version-next-backports 2>/dev/null | wc -l` -ne 0 ; then
+    git push $beQuiet -u origin +v$version-next-backports >/dev/null
```

Comment:
Missing error handling. Here it is especially bad because the script is designed to update an _existing_ branch and, hence, after the failure to push changes here, the script may create a PR with a stale set of changes -- a problem difficult for a human to notice.

Reply:
That is intentional. The preferred error handling in this script is to skip the commits that have any kind of backport issue.

Specific to `git push` failure. If *any* commits are in the PR set, that is some progress - the rest which need push'ing will hopefully work next script execution. Failing that NOC receives a notification of these script stderr outputs from both `git push` and `gh pr create`.

Reply:
You are missing two critical details:

1) if there are any changes already in `vN-next-backports` branch a PR is already open. Meaning `gh pr create` will be skipped by the current script execution.

2) if origin repository does not contain any changes after `git push` the `gh ps create` command will fail.

The result in all these cases is intentional.

Reply:
> 1. if there are any changes already in vN-next-backports branch a PR is already open

That invariant is not maintained by the proposed code AFAICT -- the branch may have changes without a corresponding PR being already open. In fact, the very next script line assumes there may be no such PR by saying "if there is an existing PR already open".

> 2) if origin repository does not contain any changes after `git push` the `gh pr create` command will fail.

The `gh pr create` command may not fail if the branch contains changes that were pushed earlier. IMHO, it is very difficult to be sure that nothing like that could ever happen in the current script version. More importantly, I see no good reason why we should spend a lot of time convincing ourselves that nothing can go wrong instead of adding simple error handling!

It looks like you have added error handling. I am resolving this old change request.